text/3604-derive_c-enum_integer_conversions.md

```diff
@@ -0,0 +1,185 @@
+```rust 
+#[derive(TryFromInt, IntoInt)]
+enum DNSOpCode {
+  //each variant is just a number, that we asociate with 
+  //one of these enum states/variants
+  StandardQuery = 0,
+  InverseQuery = 1
+  ServerStatus = 2,
+}
+```
+since the `DNSOpCode` enum maps each state to a number we can just use `as` to 
+turn it into a number, the statement below is equal to zero, because the `DNSOpCode::StandardQuery` enum variant maps to 0 (defined by our enum)
+```rust
+let opcode_num = DNSOpCode::StandardQuery as u8; //this returns 0
+```
+
+likewise, we can also use the Into trait to do the same thing
+opcode_num and opcode_num1 are the same, however opcode_num2 can be used in generic contexts where `as` cannot
+```rust
+let opcode_num2: u8 = DNSOpCode::StandardQuery.into(); //also returns 0
+```
+when going the other way, from a number to a DNSOpCode, the operation could fail, such as,
+if our user wanted to convert the number 100 to a DNSOpCode, how could we possibly convert
+that to an DNSOpCode, we can try,
+```rust 
+let try_get_opcode = DNSOpCode::try_from(100); // Err(())
+```
+well, since 100 is not one of the valid OpCodes defined in our enum, we couldn\`t convert
+the number to a DNSOpCode, lets try with another number like 2
+
+```rust
+let try_again = DNSOpCode::try_from(2); // Ok(DNSOpCode::ServerStatus)
+```
+It worked! We can construct our C-style DNSOpCode enum **safely**, with **no boilerplate**, how cool
+
+I think that this code all logically makes sense and there is really no other way I can think of to implement TryFrom<Integer> for a C-style enum, in terms of maintainability and readability I think the code isn\`t unambigous, and the reader can tell that there is one clear intent in the use of the try_form and into function, to try and get a `DNSOpCode` from an integer, or convert a `DNSOpCode` into an integer. Also, since this feature is completely additive, and it is opt-in, it won\`t break or deprecate any existing code. This feature isn\`t changing how anything new or old works, so I think it will be just as easy for beginners to learn as experienced rust programmers
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+
+I want to propose a **TryFrom<{Integer}>** and **Into<{Integer}>** derive macro becomes available for all C-style enums. C-style enums are a subset of all integers, which is why C-style Enums can be cast to integers using the ``as {usize/u8/isize, etc..}`` syntax. I propose that we create a derive macro, that automatically allows C-Style Enums to be cast to Integers using ``.into()``, not just ``as``. Since only _some_ Integers can be turned into a C-Style Enum, I propose that a TryFrom<{Integer}> derive macro can be made for them, to allow easy and more generic conversion between a C-Style Enum and Integers.
```

Comment:
> I want to propose a **TryFrom<{Integer}>** and **Into<{Integer}>** derive macro becomes available for all C-style enums.

The term "C-style enum" is ill-defined in Rust. We use either:
- "fieldless enum" for an enum where no variant has fields
- "unit-only enum" for an enum where all variants are unit-like

See the reference for more information: https://doc.rust-lang.org/reference/items/enumerations.html